crates/sui-faucet/src/faucet/simple_faucet.rs

```diff
@@ -103,57 +111,69 @@ impl SimpleFaucet {
-                                "Planning to use coin from pool {:?}, current balance: {}",
-                                coin,
-                                gas_coin.value()
-                            );
-                            return Ok(coin);
-                        } else {
-                            // If amount is not big enough, remove it from the pool
-                            self.metrics.total_discarded_coins.inc();
-                            warn!(
-                                ?uuid,
-                                "Coin {:?} does not have enough balance ({:?}), removing from pool",
-                                coin,
-                                gas_coin.value(),
-                            );
+        };
+        if let Ok(mut consumer) = consumer_res {
+            info!(?uuid, "Got consumer lock, pulling coins.");
+            loop {
+                match tokio::time::timeout(
+                    Duration::from_secs(WAIT_FOR_GAS_TIMEOUT_SEC),
+                    consumer.recv(),
+                )
+                .await
+                {
+                    Ok(Some(coin)) => {
+                        info!(?uuid, "Pulling coin from pool {:?}", coin);
+                        self.metrics.total_available_coins.dec();
+                        match self.get_gas_coin(coin).await {
+                            Ok(Some(gas_coin))
+                                if gas_coin.value() >= total_amount + PAY_SUI_GAS =>
+                            {
+                                info!(
+                                    ?uuid,
+                                    "Planning to use coin from pool {:?}, current balance: {}",
+                                    coin,
+                                    gas_coin.value()
+                                );
+                                return GasCoinResponse::ValidGasCoin(coin);
+                            }
+                            Ok(Some(_)) => {
+                                warn!(
+                                    ?uuid,
+                                    "Coin {:?} does not have sufficient balance, removing from pool", coin
+                                );
+                                return GasCoinResponse::GasCoinWithInsufficientBalance(coin);
+                            }
+                            Ok(None) => {
+                                // Invalid gas, meaning that the coin does not exist,
+                                // or does not belong to the current active address.
+                                self.metrics.total_discarded_coins.inc();
+                                warn!(
+                                    ?uuid,
+                                    "Coin {:?} is not longer valid, removing from pool", coin
+                                );
+                                return GasCoinResponse::InvalidGasCoin(coin);
+                            }
+                            Err(_) => {
```

Comment:
Here is where I mean -- we should at least print this error, if not propagate it in the return value.